### ISSUE TITLE
implement yet another plan

### ISSUE THREAD

**Issue by danregima** [ReZorg/nn.nn#9]
<div class="UserMessageLayout-module__messageContainer__IOjwG"><span class="UserProblemStatementMessage-module__messageMeta__y5RUo">You asked</span><div style="--max-height: 500px;"><div class="ExpandableMarkdown-module__overflowing__D2ZUL ExpandableMarkdown-module__expanded__H8bpC"><div class="markdown-body MarkdownRenderer-module__container__sJ8jf ExpandableMarkdown-module__content__Rn4oS UserProblemStatementMessage-module__messageText__Jng0H" data-copilot-markdown="true" style="--MarkdownRenderer_streaming-fade-duration: 750ms;"><p>Implement the plan:</p>
<h1>Implementation Plan: Comprehensive CI Build &amp; Exhaustive Unit + E2E Test Suite</h1>
<h2>Current State Assessment</h2>
<table><thead><tr><th>Component</th><th>Location</th><th>Existing tests</th><th>Runtime needed</th></tr></thead><tbody><tr><td>Torch/nn (Lua)</td><td><code node="[object Object]">lang/lua</code> (199 modules) + <code node="[object Object]">test/</code></td><td><code node="[object Object]">lang/lua/test.lua</code> (9,126 lines, <code node="[object Object]">nn.test()</code>)</td><td>Torch7 + LuaJIT (legacy, hard to build)</td></tr><tr><td>THNN C backend</td><td><code node="[object Object]">lang/c/THNN</code></td><td>none (tested via Lua suite)</td><td>CMake, TH headers</td></tr><tr><td>a9nn (pure Lua)</td><td><code node="[object Object]">lang/a9nn</code></td><td><code node="[object Object]">run_tests.lua</code> → <code node="[object Object]">test/test_basic.lua</code>, <code node="[object Object]">test/test_cognitive.lua</code>; 3 examples</td><td>Plain Lua 5.x (self-contained tensor lib)</td></tr><tr><td>Prolog</td><td><code node="[object Object]">lang/pl</code></td><td><code node="[object Object]">test_nn.pl</code>, <code node="[object Object]">test_modules.pl</code>, <code node="[object Object]">test_nd.pl</code>; 2 demos</td><td>SWI-Prolog</td></tr><tr><td>P-Lingua</td><td><code node="[object Object]">lang/pli</code></td><td><code node="[object Object]">test_nn.pli</code>, <code node="[object Object]">test_extensions.pli</code>, <code node="[object Object]">validate.sh</code> (structural checks)</td><td>bash; optional <code node="[object Object]">plingua</code> compiler</td></tr><tr><td>Raku</td><td><code node="[object Object]">lang/raku</code></td><td><code node="[object Object]">test-nn.raku</code> (TAP, <code node="[object Object]">plan 36</code>); demo/example</td><td>Rakudo</td></tr><tr><td>Racket</td><td><code node="[object Object]">lang/rkt</code></td><td><code node="[object Object]">test-nn.rkt</code>; demo/example</td><td>Racket</td></tr><tr><td>Scheme</td><td><code node="[object Object]">lang/scm</code></td><td><code node="[object Object]">test-nn.scm</code>; demo/example</td><td>Guile</td></tr><tr><td>Docs</td><td><code node="[object Object]">doc/</code>, <code node="[object Object]">docs/</code>, <code node="[object Object]">mkdocs.yml</code></td><td>none</td><td>mkdocs</td></tr><tr><td>CI</td><td><code node="[object Object]">.travis.yml</code> only</td><td>—</td><td>Travis is defunct</td></tr></tbody></table>
<p>Key gaps: no GitHub Actions, no unified test entrypoint, several test suites don't set a nonzero exit code on failure, no cross-implementation consistency checks, no coverage of <code node="[object Object]">lang/lua</code> core modules without a full Torch build.</p>
<h2>Phase 1: CI Skeleton &amp; Change Detection</h2>
<ol>
<li node="[object Object]">Create <code node="[object Object]">.github/workflows/ci.yml</code> as the main workflow, triggered on <code node="[object Object]">push</code> (default branch) and <code node="[object Object]">pull_request</code>.</li>
<li node="[object Object]">Add a <code node="[object Object]">changes</code> job using path filtering (e.g., <code node="[object Object]">dorny/paths-filter</code>) so each language job only runs when its <code node="[object Object]">lang/&lt;x&gt;/**</code> subtree (or shared files) changes; run everything on the default branch and on manual <code node="[object Object]">workflow_dispatch</code>.</li>
<li node="[object Object]">Add repo-wide fast checks that always run: Luacheck lint (using existing <code node="[object Object]">.luacheckrc</code>) over <code node="[object Object]">lang/lua</code> and <code node="[object Object]">lang/a9nn</code>, plus a markdown/link sanity job for the extensive docs.</li>
<li node="[object Object]">Deprecate/remove <code node="[object Object]">.travis.yml</code> (or keep it clearly marked legacy) and update the README build badge to point at GitHub Actions.</li>
</ol>
<h2>Phase 2: Per-Language Build &amp; Unit-Test Jobs</h2>
<p>Each job installs its toolchain via apt/setup actions, then runs the existing suite. Harden every runner so failures produce nonzero exit codes:</p>
<ol>
<li node="[object Object]"><strong>a9nn (Lua)</strong>: matrix over Lua 5.1/5.3/5.4 + LuaJIT; run <code node="[object Object]">lang/a9nn/run_tests.lua</code>. Fix the runner to exit nonzero on any suite failure and to aggregate pass/fail counts.</li>
<li node="[object Object]"><strong>Prolog</strong>: SWI-Prolog; run all three test files with <code node="[object Object]">swipl -q -l &lt;file&gt; -g &lt;goal&gt; -t halt</code>; ensure failing tests cause <code node="[object Object]">halt(1)</code>.</li>
<li node="[object Object]"><strong>Raku</strong>: Rakudo (via <code node="[object Object]">Raku/setup-raku</code> or rakudo-pkg); run <code node="[object Object]">raku test-nn.raku</code> under <code node="[object Object]">prove6</code>/TAP so plan mismatches fail the job.</li>
<li node="[object Object]"><strong>Racket</strong>: <code node="[object Object]">Bogdanp/setup-racket</code>; run <code node="[object Object]">racket test-nn.rkt</code>; add an exit-code check based on the fail counter (currently it only prints).</li>
<li node="[object Object]"><strong>Scheme</strong>: Guile via apt; run <code node="[object Object]">guile -l nn.scm -l test-nn.scm</code> with an added final form that exits nonzero if <code node="[object Object]">test-failed &gt; 0</code>. Verify Guile compatibility (the file mentions MIT Scheme <code node="[object Object]">format</code> usage) and adjust invocation or add a small compatibility shim if needed.</li>
<li node="[object Object]"><strong>P-Lingua</strong>: run <code node="[object Object]">lang/pli/validate.sh</code> (structural checks, no dependencies); optionally add a second step that installs the <code node="[object Object]">plingua</code> compiler in a container to compile all <code node="[object Object]">.pli</code> files, marked <code node="[object Object]">continue-on-error</code> initially.</li>
<li node="[object Object]"><strong>THNN C build</strong>: a compile-only smoke job (gcc + clang matrix) that builds <code node="[object Object]">lang/c/THNN</code> against Torch7 headers, or at minimum runs a header/syntax check (<code node="[object Object]">gcc -fsyntax-only</code>) if a full Torch checkout proves too fragile — decide during implementation.</li>
<li node="[object Object]"><strong>Torch/nn full suite (legacy)</strong>: recreate the Travis pipeline in a dedicated, <code node="[object Object]">continue-on-error</code>, cache-heavy job (OpenBLAS cache, torch/distro build, <code node="[object Object]">luarocks make rocks/nn-scm-1.rockspec</code>, then <code node="[object Object]">nn.test()</code>). Given Torch7's age, gate this behind a nightly <code node="[object Object]">schedule</code> + manual dispatch rather than every PR, and document that it's best-effort.</li>
</ol>
<h2>Phase 3: Exhaustive Unit-Test Expansion</h2>
<ol>
<li node="[object Object]"><strong>a9nn</strong>: extend <code node="[object Object]">test/test_basic.lua</code> and <code node="[object Object]">test/test_cognitive.lua</code> for untested modules (EchoReservoir, AtomSpace, Personality, EpisodicMemory, LLaMAOrchestrator stubs, NNECCOAgent), including edge cases (empty inputs, dimension mismatches, serialization round-trips).</li>
<li node="[object Object]"><strong>Prolog</strong>: add coverage for any predicates in <code node="[object Object]">nn.pl</code>/<code node="[object Object]">nn_nd.pl</code> not exercised (fractal modules, criterion gradients); assert numeric tolerances rather than just success.</li>
<li node="[object Object]"><strong>Scheme/Racket/Raku</strong>: mirror each README's feature list against the test files and fill gaps (dropout, batch norm if present, containers, all activations/criterions), plus numeric gradient-checking tests (finite-difference vs. analytic backprop) — the strongest correctness signal for NN code.</li>
<li node="[object Object]"><strong>Cross-implementation consistency tests</strong>: a small shared fixture (JSON/CSV of fixed weights, an input, and expected forward output + loss) checked by each language's suite, guaranteeing all seven ports agree to a tolerance.</li>
</ol>
<h2>Phase 4: E2E Test Suite</h2>
<ol>
<li node="[object Object]">Treat each language's <code node="[object Object]">demo</code>/<code node="[object Object]">example</code> programs as e2e tests: run them in CI with a fixed random seed where supported and assert successful exit plus a convergence criterion (e.g., XOR training loss below threshold), rather than only "didn't crash".</li>
<li node="[object Object]">a9nn e2e: run <code node="[object Object]">examples/mlp_example.lua</code>, <code node="[object Object]">multi_agent_example.lua</code>, <code node="[object Object]">nnecco_example.lua</code>; where an example depends on external llama.cpp instances, add a mock/offline mode so it runs deterministically in CI.</li>
<li node="[object Object]">Add timeouts per e2e step and mark inherently nondeterministic demos with retry-once semantics.</li>
</ol>
<h2>Phase 5: Aggregation, Docs &amp; Quality Gates</h2>
<ol>
<li node="[object Object]">Add a final <code node="[object Object]">ci-success</code> fan-in job that requires all matrix jobs, suitable as a single required status check for branch protection.</li>
<li node="[object Object]">Add a <code node="[object Object]">docs</code> job: <code node="[object Object]">mkdocs build --strict</code> to catch broken nav/links in <code node="[object Object]">mkdocs.yml</code>, <code node="[object Object]">doc/</code>, <code node="[object Object]">docs/</code>.</li>
<li node="[object Object]">Upload test logs/TAP output as artifacts; publish a summary table via <code node="[object Object]">$GITHUB_STEP_SUMMARY</code> per job.</li>
<li node="[object Object]">Cache toolchains (Rakudo, Racket, OpenBLAS/Torch build) to keep PR CI fast.</li>
<li node="[object Object]">Update <code node="[object Object]">README.md</code>, <code node="[object Object]">CONTRIBUTING.md</code>, and each <code node="[object Object]">lang/*/README.md</code> with the canonical local test commands matching CI exactly.</li>
</ol>
<h2>Suggested Delivery Order</h2>
<ol>
<li node="[object Object]">CI skeleton + lint + the five cheap language jobs (a9nn, Prolog, Raku, Racket, Scheme, P-Lingua structural) — immediate signal with existing tests.</li>
<li node="[object Object]">Exit-code hardening of the test runners (small code fixes so CI can actually fail).</li>
<li node="[object Object]">E2E jobs from demos/examples.</li>
<li node="[object Object]">Unit-test expansion + cross-implementation consistency fixtures.</li>
<li node="[object Object]">THNN compile job and the nightly legacy Torch full-suite job.</li>
<li node="[object Object]">Docs job, fan-in gate, badges, and documentation updates.</li>
</ol></div></div><button data-component="Button" type="button" aria-expanded="true" class="prc-Button-ButtonBase-9n-Xk ExpandableMarkdown-module__toggle__NBxAX" data-loading="false" data-no-visuals="true" data-size="small" data-variant="invisible"><span data-component="buttonContent" data-align="center" class="prc-Button-ButtonContent-Iohp5"><span data-component="text" class="prc-Button-Label-FWkx3">Show less</span></span></button></div></div>
